hadoop-hdds/server-scm/src/main/java/org/apache/hadoop/hdds/scm/container/ContainerManagerImpl.java

```diff
@@ -0,0 +1,282 @@
+        } catch (ContainerNotFoundException e) {
+          // How can this happen? o_O
+          return null;
+        }
+      }).filter(Objects::nonNull).collect(Collectors.toSet());
+    } finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  @Override
+  public boolean exists(final ContainerID containerID) {
+    lock.readLock().lock();
+    try {
+      return (containerStateManager.getContainer(containerID) != null);
+    } catch (ContainerNotFoundException ex) {
+      return false;
+    } finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  @Override
+  public List<ContainerInfo> listContainers(final ContainerID startID,
+                                            final int count) {
+    lock.readLock().lock();
+    try {
+      final long startId = startID == null ? 0 : startID.getId();
+      final List<ContainerID> containersIds =
+          new ArrayList<>(containerStateManager.getContainerIDs());
+      Collections.sort(containersIds);
+      return containersIds.stream()
+          .filter(id -> id.getId() > startId)
+          .limit(count)
+          .map(id -> {
+            try {
+              return containerStateManager.getContainer(id);
+            } catch (ContainerNotFoundException ex) {
+              // This can never happen, as we hold lock no one else can remove
+              // the container after we got the container ids.
+              LOG.warn("Container Missing.", ex);
+              return null;
+            }
+          }).collect(Collectors.toList());
+    } finally {
+      lock.readLock().unlock();
+    }
+  }
+
+  @Override
+  public ContainerInfo allocateContainer(final ReplicationType type,
+      final ReplicationFactor replicationFactor, final String owner)
+      throws IOException {
+    lock.writeLock().lock();
+    try {
+      final List<Pipeline> pipelines = pipelineManager
+          .getPipelines(type, replicationFactor, Pipeline.PipelineState.OPEN);
+
+      if (pipelines.isEmpty()) {
+        throw new IOException("Could not allocate container. Cannot get any" +
+            " matching pipeline for Type:" + type + ", Factor:" +
+            replicationFactor + ", State:PipelineState.OPEN");
+      }
+
+      final ContainerID containerID = containerStateManager
+          .getNextContainerID();
+      final Pipeline pipeline = pipelines.get(
+          (int) containerID.getId() % pipelines.size());
+
+      final ContainerInfoProto containerInfo = ContainerInfoProto.newBuilder()
+          .setState(LifeCycleState.OPEN)
+          .setPipelineID(pipeline.getId().getProtobuf())
+          .setUsedBytes(0)
+          .setNumberOfKeys(0)
+          .setStateEnterTime(Time.now())
+          .setOwner(owner)
+          .setContainerID(containerID.getId())
+          .setDeleteTransactionId(0)
+          .setReplicationFactor(pipeline.getFactor())
+          .setReplicationType(pipeline.getType())
+          .build();
+      containerStateManager.addContainer(containerInfo);
```

Comment:
Do we need to addContainerToDB here?

Reply:
It is handled inside `ContainerStateManager`.